main/src/ca/uwaterloo/flix/language/phase/Mutator.scala

```diff
@@ -0,0 +1,149 @@
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.Ast.Constant
+import ca.uwaterloo.flix.language.ast.SemanticOp.BoolOp
+import ca.uwaterloo.flix.language.ast.TypedAst._
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation._
+
+object Mutator {
+
+  // TODO: The purpose of this phase is to:
+  // 1. Pick a random def (perhaps rather a def from a specific file ?)
+  // 2. Pick a random expr (how?)
+  // 3. Slightly modify the expr (e.g. changing (x + 1) to (x - 1).
+
+  // Once this phase has been implemented, then we can introduce a new MutationTester class and a few changes to the Flix object to tie everything together.
+
+  def run(root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = flix.phase("Mutator") {
+
+    // Pick a random def
+    val pickedDef = random(root.defs.values.toList)
+
+    // Mutate the def.
+    val mutatedDef = mutateDef(pickedDef)
+
+    // Rebuild the AST and return it.
+    root.copy(defs = root.defs + (mutatedDef.sym -> mutatedDef)).toSuccess
+  }
+
+  private def mutateDef(defn: Def): Def = {
+    // We mutate the expression and rebuild in the def.
+    val mutatedExp = mutateExpr(defn.impl.exp)
+
+    defn.copy(impl = defn.impl.copy(exp = mutatedExp))
+  }
+
+  // TODO recurse on the expr, find a random node, and mutate it.
+  private def mutateExpr(exp: Expr): Expr = exp match {
+    case Expr.Cst(cst, tpe, loc) => cst match {
+      case Constant.Unit => exp
+      case Constant.Null => ???
+      case Constant.Bool(lit) => ???
+      case Constant.Char(lit) => ???
+      case Constant.Float32(lit) => ???
+      case Constant.Float64(lit) => ???
+      case Constant.BigDecimal(lit) => ???
+      case Constant.Int8(lit) => ???
+      case Constant.Int16(lit) => ???
+      case Constant.Int32(lit) =>
```

Comment:
Do we want to be able to create more than 1 mutations if necessary? I think that would be a good solution. For example, for Int n : `n + 1` and ` n - 1`.

Reply:
Usually one makes exactly one change to the program.

Reply:
Is it right in this case to create separate mutators for different changes? For example, to increase or decrease a number by 1, it can be MutatorIncrementer and MutatorDecrementer.